File: backend/routes/gst_ledger.py
from fastapi import APIRouter, HTTPException, Request
from database import db
from datetime import datetime, timezone
from typing import Optional
import uuid

router = APIRouter()

# ============ GST LEDGER & OPENING BALANCE ============

@router.get("/gst-ledger/opening-balance")
async def get_gst_opening_balance(kms_year: str):
    saved = await db.gst_opening_balances.find_one({"kms_year": kms_year}, {"_id": 0})
    if saved:
        return {"igst": saved.get("igst", 0), "sgst": saved.get("sgst", 0), "cgst": saved.get("cgst", 0), "kms_year": kms_year, "source": "manual"}
    return {"igst": 0, "sgst": 0, "cgst": 0, "kms_year": kms_year, "source": "none"}

@router.put("/gst-ledger/opening-balance")
async def save_gst_opening_balance(request: Request):
    data = await request.json()
    kms_year = data.get("kms_year")
    if not kms_year:
        raise HTTPException(status_code=400, detail="kms_year is required")
    doc = {
        "kms_year": kms_year,
        "igst": float(data.get("igst", 0)),
        "sgst": float(data.get("sgst", 0)),
        "cgst": float(data.get("cgst", 0)),
        "updated_at": datetime.now(timezone.utc).isoformat()
    }
    await db.gst_opening_balances.update_one({"kms_year": kms_year}, {"$set": doc}, upsert=True)
    return doc


@router.get("/gst-ledger")
async def get_gst_ledger(kms_year: Optional[str] = None, season: Optional[str] = None):
    """Compute GST ledger from all vouchers (purchase = credit/add, sale = debit/minus)"""
    query = {}
    if kms_year: query["kms_year"] = kms_year
    if season: query["season"] = season

    # Get opening balance
    ob = {"igst": 0, "sgst": 0, "cgst": 0}
    if kms_year:
        saved_ob = await db.gst_opening_balances.find_one({"kms_year": kms_year}, {"_id": 0})
        if saved_ob:
            ob = {"igst": saved_ob.get("igst", 0), "sgst": saved_ob.get("sgst", 0), "cgst": saved_ob.get("cgst", 0)}

    entries = []

    # Purchase vouchers → GST credit (add)
    purchases = await db.purchase_vouchers.find(query, {"_id": 0}).sort("date", 1).to_list(5000)
    for p in purchases:
        cgst = p.get("cgst_amount", 0) or 0
        sgst = p.get("sgst_amount", 0) or 0
        igst = p.get("igst_amount", 0) or 0
        if cgst > 0 or sgst > 0 or igst > 0:
            entries.append({
                "date": p.get("date", ""), "type": "purchase", "voucher_type": "Purchase",
                "voucher_no": p.get("voucher_no", ""), "party": p.get("party_name", ""),
                "description": f"Purchase #{p.get('voucher_no','')} - {p.get('party_name','')}",
                "cgst": round(cgst, 2), "sgst": round(sgst, 2), "igst": round(igst, 2),
                "direction": "credit", "id": p.get("id", "")
            })

    # Gunny Bag purchases → GST credit (add)
    gunnys = await db.gunny_bags.find({**query, "txn_type": "in"}, {"_id": 0}).sort("date", 1).to_list(5000)
    for g in gunnys:
        cgst = g.get("cgst_amount", 0) or 0
        sgst = g.get("sgst_amount", 0) or 0
        igst = g.get("gst_amount", 0) or 0  # igst stored as gst_amount in gunny
        if g.get("gst_type") == "igst":
            igst = g.get("gst_amount", 0) or 0
            cgst = 0; sgst = 0
        elif g.get("gst_type") == "cgst_sgst":
            igst = 0
        if cgst > 0 or sgst > 0 or igst > 0:
            entries.append({
                "date": g.get("date", ""), "type": "purchase", "voucher_type": "Gunny Bag",
                "voucher_no": g.get("invoice_no", ""), "party": g.get("party_name", g.get("source", "")),
                "description": f"Gunny Bag - {g.get('party_name', g.get('source',''))}",
                "cgst": round(cgst, 2), "sgst": round(sgst, 2), "igst": round(igst, 2),
                "direction": "credit", "id": g.get("id", "")
            })

    # Sale vouchers → GST debit (minus)
    sales = await db.sale_vouchers.find(query, {"_id": 0}).sort("date", 1).to_list(5000)
    for s in sales:
        cgst = s.get("cgst_amount", 0) or 0
        sgst = s.get("sgst_amount", 0) or 0
        igst = s.get("igst_amount", 0) or 0
        if cgst > 0 or sgst > 0 or igst > 0:
            entries.append({
                "date": s.get("date", ""), "type": "sale", "voucher_type": "Sale",
                "voucher_no": s.get("voucher_no", ""), "party": s.get("party_name", ""),
                "description": f"Sale #{s.get('voucher_no','')} - {s.get('party_name','')}",
                "cgst": round(cgst, 2), "sgst": round(sgst, 2), "igst": round(igst, 2),
                "direction": "debit", "id": s.get("id", "")
            })

    # Sort all entries by date
    entries.sort(key=lambda x: x.get("date", ""))

    # Calculate running balance
    running_cgst = ob["cgst"]
    running_sgst = ob["sgst"]
    running_igst = ob["igst"]
    for e in entries:
        if e["direction"] == "credit":
            running_cgst += e["cgst"]; running_sgst += e["sgst"]; running_igst += e["igst"]
        else:
            running_cgst -= e["cgst"]; running_sgst -= e["sgst"]; running_igst -= e["igst"]
        e["running_cgst"] = round(running_cgst, 2)
        e["running_sgst"] = round(running_sgst, 2)
        e["running_igst"] = round(running_igst, 2)

    # Summary
    total_credit_cgst = round(sum(e["cgst"] for e in entries if e["direction"] == "credit"), 2)
    total_credit_sgst = round(sum(e["sgst"] for e in entries if e["direction"] == "credit"), 2)
    total_credit_igst = round(sum(e["igst"] for e in entries if e["direction"] == "credit"), 2)
    total_debit_cgst = round(sum(e["cgst"] for e in entries if e["direction"] == "debit"), 2)
    total_debit_sgst = round(sum(e["sgst"] for e in entries if e["direction"] == "debit"), 2)
    total_debit_igst = round(sum(e["igst"] for e in entries if e["direction"] == "debit"), 2)

    return {
        "opening_balance": ob,
        "entries": entries,
        "summary": {
            "credit": {"cgst": total_credit_cgst, "sgst": total_credit_sgst, "igst": total_credit_igst},
            "debit": {"cgst": total_debit_cgst, "sgst": total_debit_sgst, "igst": total_debit_igst},
            "balance": {
                "cgst": round(ob["cgst"] + total_credit_cgst - total_debit_cgst, 2),
                "sgst": round(ob["sgst"] + total_credit_sgst - total_debit_sgst, 2),
                "igst": round(ob["igst"] + total_credit_igst - total_debit_igst, 2),
            }
        },
        "total_entries": len(entries)
    }


@router.get("/govt-bags/stock")
async def get_govt_bags_stock(kms_year: Optional[str] = None, season: Optional[str] = None):
    """Get govt (free/new) bags stock summary"""
    query = {"bag_type": "new"}
    if kms_year: query["kms_year"] = kms_year
    if season: query["season"] = season
    bags = await db.gunny_bags.find(query, {"_id": 0}).to_list(5000)
    bags_in = sum(b.get("quantity", 0) for b in bags if b.get("txn_type") == "in")
    bags_out = sum(b.get("quantity", 0) for b in bags if b.get("txn_type") == "out")
    return {"bags_in": bags_in, "bags_out": bags_out, "stock": bags_in - bags_out}

File: desktop-app/routes/gst_ledger.js
const express = require('express');
const { safeHandler } = require('./safe_handler');
const { fmtDate } = require('./pdf_helpers');

module.exports = function(database) {
  const router = express.Router();

  // GET /api/gst-ledger/opening-balance
  router.get('/api/gst-ledger/opening-balance', safeHandler(async (req, res) => {
    const kms_year = req.query.kms_year;
    if (!database.data.gst_opening_balances) database.data.gst_opening_balances = {};
    const saved = database.data.gst_opening_balances[kms_year];
    if (saved) return res.json({ igst: saved.igst || 0, sgst: saved.sgst || 0, cgst: saved.cgst || 0, kms_year, source: 'manual' });
    res.json({ igst: 0, sgst: 0, cgst: 0, kms_year, source: 'none' });
  }));

  // PUT /api/gst-ledger/opening-balance
  router.put('/api/gst-ledger/opening-balance', safeHandler(async (req, res) => {
    const { kms_year, igst, sgst, cgst } = req.body;
    if (!kms_year) return res.status(400).json({ detail: 'kms_year is required' });
    if (!database.data.gst_opening_balances) database.data.gst_opening_balances = {};
    const doc = { kms_year, igst: parseFloat(igst) || 0, sgst: parseFloat(sgst) || 0, cgst: parseFloat(cgst) || 0, updated_at: new Date().toISOString() };
    database.data.gst_opening_balances[kms_year] = doc;
    database.save();
    res.json(doc);
  }));

  // GET /api/gst-ledger
  router.get('/api/gst-ledger', safeHandler(async (req, res) => {
    const { kms_year, season } = req.query;
    if (!database.data.gst_opening_balances) database.data.gst_opening_balances = {};
    const ob = database.data.gst_opening_balances[kms_year] || { igst: 0, sgst: 0, cgst: 0 };

    const entries = [];
    const filter = (arr) => (arr || []).filter(v => (!kms_year || v.kms_year === kms_year) && (!season || v.season === season));

    // Purchase vouchers → credit
    filter(database.data.purchase_vouchers).forEach(p => {
      const cgst = p.cgst_amount || 0, sgst = p.sgst_amount || 0, igst = p.igst_amount || 0;
      if (cgst > 0 || sgst > 0 || igst > 0) {
        entries.push({ date: p.date || '', type: 'purchase', voucher_type: 'Purchase', voucher_no: p.voucher_no || '', party: p.party_name || '',
          description: `Purchase #${p.voucher_no || ''} - ${p.party_name || ''}`, cgst: Math.round(cgst * 100) / 100, sgst: Math.round(sgst * 100) / 100,
          igst: Math.round(igst * 100) / 100, direction: 'credit', id: p.id || '' });
      }
    });

    // Gunny bags purchases → credit
    filter(database.data.gunny_bags).filter(g => g.txn_type === 'in').forEach(g => {
      let cgst = g.cgst_amount || 0, sgst = g.sgst_amount || 0, igst = 0;
      if (g.gst_type === 'igst') { igst = g.gst_amount || 0; cgst = 0; sgst = 0; }
      if (cgst > 0 || sgst > 0 || igst > 0) {
        entries.push({ date: g.date || '', type: 'purchase', voucher_type: 'Gunny Bag', voucher_no: g.invoice_no || '', party: g.party_name || g.source || '',
          description: `Gunny Bag - ${g.party_name || g.source || ''}`, cgst: Math.round(cgst * 100) / 100, sgst: Math.round(sgst * 100) / 100,
          igst: Math.round(igst * 100) / 100, direction: 'credit', id: g.id || '' });
      }
    });

    // Sale vouchers → debit
    filter(database.data.sale_vouchers).forEach(s => {
      const cgst = s.cgst_amount || 0, sgst = s.sgst_amount || 0, igst = s.igst_amount || 0;
      if (cgst > 0 || sgst > 0 || igst > 0) {
        entries.push({ date: s.date || '', type: 'sale', voucher_type: 'Sale', voucher_no: s.voucher_no || '', party: s.party_name || '',
          description: `Sale #${s.voucher_no || ''} - ${s.party_name || ''}`, cgst: Math.round(cgst * 100) / 100, sgst: Math.round(sgst * 100) / 100,
          igst: Math.round(igst * 100) / 100, direction: 'debit', id: s.id || '' });
      }
    });

    entries.sort((a, b) => (a.date || '').slice(0,10).localeCompare((b.date || '').slice(0,10)));

    let rc = ob.cgst || 0, rs = ob.sgst || 0, ri = ob.igst || 0;
    for (const e of entries) {
      if (e.direction === 'credit') { rc += e.cgst; rs += e.sgst; ri += e.igst; }
      else { rc -= e.cgst; rs -= e.sgst; ri -= e.igst; }
      e.running_cgst = Math.round(rc * 100) / 100;
      e.running_sgst = Math.round(rs * 100) / 100;
      e.running_igst = Math.round(ri * 100) / 100;
    }

    const sum = (dir, field) => Math.round(entries.filter(e => e.direction === dir).reduce((s, e) => s + (e[field] || 0), 0) * 100) / 100;
    res.json({
      opening_balance: { igst: ob.igst || 0, sgst: ob.sgst || 0, cgst: ob.cgst || 0 },
      entries,
      summary: {
        credit: { cgst: sum('credit', 'cgst'), sgst: sum('credit', 'sgst'), igst: sum('credit', 'igst') },
        debit: { cgst: sum('debit', 'cgst'), sgst: sum('debit', 'sgst'), igst: sum('debit', 'igst') },
        balance: { cgst: Math.round(rc * 100) / 100, sgst: Math.round(rs * 100) / 100, igst: Math.round(ri * 100) / 100 }
      },
      total_entries: entries.length
    });
  }));

  // GET /api/govt-bags/stock
  router.get('/api/govt-bags/stock', safeHandler(async (req, res) => {
    const { kms_year, season } = req.query;
    const bags = (database.data.gunny_bags || []).filter(b => b.bag_type === 'new' && (!kms_year || b.kms_year === kms_year) && (!season || b.season === season));
    const bags_in = bags.filter(b => b.txn_type === 'in').reduce((s, b) => s + (b.quantity || 0), 0);
    const bags_out = bags.filter(b => b.txn_type === 'out').reduce((s, b) => s + (b.quantity || 0), 0);
    res.json({ bags_in, bags_out, stock: bags_in - bags_out });
  }));

  return router;
};
